src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -3377,6 +3377,29 @@ TransactionStateMachine(WorkerSession *session)
 			case REMOTE_TRANS_SENT_COMMAND:
 			{
 				TaskPlacementExecution *placementExecution = session->currentTask;
+				if (placementExecution == NULL)
+				{
+					/*
+					 * We have seen accounts in production where the placementExecution
+					 * could inadvertently be not set. Investigation documented on
+					 * https://github.com/citusdata/citus-enterprise/issues/493
+					 * (due to sensitive data in the initial report it is not discussed
+					 * in our community repository)
+					 *
+					 * Currently we don't have a reliable way of reproducing this issue.
+					 * Erroring here seems to be a more desirable approach compared to a
+					 * SEGFAULT on the dereference of placementExecution, with a possible
+					 * crash recovery as a result.
+					 */
+					ereport(ERROR, (errmsg(
+										"unable to recover from inconsistent state in "
+										"the connection state machine on coordinator"),
+									errhint(
+										"it is advised to close the current session and "
+										"start a new sessions to make sure a full reset "
+										"of the internal state can be performed")));
```

Comment:
if this is the advice, should this be FATAL? (which kills the session)

Reply:
Makes a lot of sense I think.
I don't know if it is required to restart the session though.

Since it is not reproducible on demand it is hard to verify if the session is still usable after this is reached at the moment.

Reply:
As we talked with @thanodnl in person, and had the opportunity to test on the repro, the user is not required to re-connect. It is mainly because all the transaction state is cleared and the new transaction would not have any problems regarding the weird state.

So, we thought we can remove the HINT and avoid `FATAL`?